components/scim/org.wso2.carbon.identity.scim.provider/src/main/java/org/wso2/carbon/identity/scim/provider/impl/ProvisioningEntityBuilder.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.scim.provider.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.common.IdentityApplicationManagementException;
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
+import org.wso2.carbon.identity.provisioning.IdentityProvisioningConstants;
+import org.wso2.carbon.identity.provisioning.ProvisioningEntity;
+import org.wso2.carbon.identity.provisioning.ProvisioningEntityType;
+import org.wso2.carbon.identity.provisioning.ProvisioningOperation;
+import org.wso2.carbon.identity.provisioning.dao.ProvisioningManagementDAO;
+import org.wso2.carbon.identity.scim.common.utils.AttributeMapper;
+import org.wso2.carbon.user.core.util.UserCoreUtil;
+import org.wso2.charon.core.attributes.SimpleAttribute;
+import org.wso2.charon.core.exceptions.CharonException;
+import org.wso2.charon.core.exceptions.NotFoundException;
+import org.wso2.charon.core.objects.AbstractSCIMObject;
+import org.wso2.charon.core.objects.Group;
+import org.wso2.charon.core.objects.SCIMObject;
+import org.wso2.charon.core.objects.User;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public final class ProvisioningEntityBuilder {
+
+    private static volatile ProvisioningEntityBuilder provisioningEntityBuilder = null;
+    private ProvisioningManagementDAO provisioningManagementDAO = null;
+    private static Log log = LogFactory.getLog(ProvisioningEntityBuilder.class);
+
+    private ProvisioningEntityBuilder() {
+        provisioningManagementDAO = new ProvisioningManagementDAO();
+    }
+
+    public static ProvisioningEntityBuilder getInstance() {
+        if (provisioningEntityBuilder == null) {
+            synchronized (ProvisioningEntityBuilder.class) {
+                provisioningEntityBuilder = new ProvisioningEntityBuilder();
+            }
+        }
+        return provisioningEntityBuilder;
+    }
+
+    ProvisioningEntity getProvisioningEntityForUserAdd(SCIMObject provisioningObject,
```

Comment:
Shall we rename the methods to buildXXX instead of getXXX since the class is also called Builder ?